c/embeddedjs.c

```diff
@@ -0,0 +1,710 @@
+                      int *statusPtr){
+  JSContext *ctx = ejs->ctx;
+  JSValue val;
+  int ret;
+  
+  if ((eval_flags & JS_EVAL_TYPE_MASK) == JS_EVAL_TYPE_MODULE) {
+    /* for the modules, we compile then run to be able to set
+       import.meta */
+    val = JS_Eval(ctx, buffer, bufferLength, filename,
+                  eval_flags | JS_EVAL_FLAG_COMPILE_ONLY);
+    if (!JS_IsException(val)) {
+      js_module_set_import_meta(ctx, val, TRUE, TRUE);
+      val = JS_EvalFunction(ctx, val);
+    }
+  } else {
+    val = JS_Eval(ctx, buffer, bufferLength, filename, eval_flags);
+  }
+  if (JS_IsException(val)) {
+      js_std_dump_error(ctx);
+      ret = -1;
+  } else {
+    ret = 0;
+  }
+  *statusPtr = ret;
+  return val;
+}
+
+void ejsFreeJSValue(EmbeddedJS *ejs, JSValue value){
+  JS_FreeValue(ejs->ctx, value);
+}
+
+int ejsSetGlobalProperty(EmbeddedJS *ejs, const char *propertyName, JSValue value){
+  JSContext *ctx = ejs->ctx;
+  JSValue theGlobal = JS_GetGlobalObject(ctx);
+
+  return JS_SetPropertyStr(ctx,theGlobal,propertyName,value);
+}
+
+int ejsEvalFile(EmbeddedJS *ejs, const char *filename, int loadMode){
+    uint8_t *buf;
+    int ret, eval_flags;
+    size_t bufferLength;
+    
+    buf = js_load_file(ejs->ctx, &bufferLength, filename);
+    if (!buf) {
+        perror(filename);
+        exit(1);
+    }
+
+    /* Beware of overly-expedient 3-valued logic here! */
+    if (loadMode < 0) {
+        loadMode = (has_suffix(filename, ".mjs") ||
+                  JS_DetectModule((const char *)buf, bufferLength));
+    }
+    if (loadMode)
+        eval_flags = JS_EVAL_TYPE_MODULE;
+    else
+        eval_flags = JS_EVAL_TYPE_GLOBAL;
+    JSValue evalResult = ejsEvalBuffer(ejs, buf, bufferLength, filename, eval_flags, &ret);
+    js_free(ejs->ctx, buf);
+    JS_FreeValue(ejs->ctx, evalResult);
+    return ret;
+}
+
+static JSClassID js_joe_thingy_class_id;
+
+static void js_joe_thingy_finalizer(JSRuntime *rt, JSValue val){
+  /* JSSTDFile *s = JS_GetOpaque(val, js_std_file_class_id); */
+  printf("Joe Thingy Finalizer running\n");
+}
+
+
+static JSClassDef js_joe_thingy_class = {
+    "THINGY",
+    .finalizer = js_joe_thingy_finalizer,
+};
+
+/*
+  #define JS_CFUNC_DEF(name, length, func1) 
+
+  name is JS name
+
+  #define JS_CFUNC_MAGIC_DEF(name, length, func1, magic)
+
+*/
+
+JSValue js_joe_thingy_three(JSContext *ctx, JSValueConst this_val, int argc, JSValueConst *argv){
+  return JS_NewInt32(ctx, 3);
+}
+
+static const JSCFunctionListEntry js_joe_thingy_proto_funcs[] = {
+    JS_CFUNC_DEF("three", 0, js_joe_thingy_three ),
+};
+
+static JSValue js_joe_boop(JSContext *ctx, JSValueConst this_val,
+                           int argc, JSValueConst *argv)
+{
+  printf("boop boop boop!!\n");
+    return JS_UNDEFINED;
+}
+
+static const JSCFunctionListEntry js_joe_funcs[] = {
+    JS_CFUNC_DEF("boop", 0, js_joe_boop ),
+};
+
+
+
+static int js_joe_init(JSContext *ctx, JSModuleDef *m)
```

Comment:
Not sure if we should mention names in code. At least there was a rule to remove that when we were open-sourcing.

Reply:
Fixed.